File: src/spoj/dynamic_programming/ANARC05H.java
package spoj.dynamic_programming;

// http://www.spoj.com/problems/ANARC05H

public class ANARC05H {

	// TODO: use cumulative sum array
	static long sum(int left, int right, int[] digits) {
		long sum = 0;
		for (int i = left; i <= right; i++) {
			sum += digits[i];
		}
		return sum;
	}

	// Exponential complexity
	// TODO: use memoization
	static long variants(int left, long prev, int[] digits) {
		if (left == digits.length) {
			return 1;
		}

		long result = 0;

		for (int k = left; k < digits.length; k++) {
			long preffixSum = sum(left, k, digits);

			if (preffixSum >= prev) {
				result += variants(k + 1, preffixSum, digits);
			}
		}

		return result;
	}

	static long variants(int[] digits) {
		return variants(0, 0, digits);
	}

	public static void main(String[] args) {
		System.out.println(variants(new int[]{6, 3, 5}));
		System.out.println(variants(new int[]{1, 1, 1, 7}));
		System.out.println(variants(new int[]{9, 8, 7, 6}));
		System.out.println(variants(new int[]{0, 0, 0}));
		System.out.println(variants(new int[]{0, 0, 0, 0}));
		System.out.println(variants(new int[]{0, 0, 0, 0, 0}));
	}
}

File: src/spoj/dynamic_programming/ANARC05H.c
// http://www.spoj.com/problems/ANARC05H/

#include <stdio.h>
#include <string.h>

#define MAX_SIZE 26

char DIGITS[MAX_SIZE];

long CUMULATIVE_SUM[MAX_SIZE + 1];

long MEMOIZED_VALUES[MAX_SIZE][MAX_SIZE * 9];

int IS_MEMOIZED[MAX_SIZE][MAX_SIZE * 9];

void calculateCumulativeSum(int length) {
    int i;
    
    CUMULATIVE_SUM[0] = 0;
    for(i = 1; i <= length; i++) {
        CUMULATIVE_SUM[i] = CUMULATIVE_SUM[i - 1] + (DIGITS[i - 1] - '0');
    }
}

long calculateVariants(int left, int prevSum, int length) {
    if(left >= length) {
        return 1;
    }
    
    if(IS_MEMOIZED[left][prevSum]) {
        return MEMOIZED_VALUES[left][prevSum];
    }
    
    long variants = 0;
    
    int k;
    long preffixSum;
    
    for(k = left; k < length; k++) {
        preffixSum = CUMULATIVE_SUM[k + 1] - CUMULATIVE_SUM[left];
        
        if(preffixSum >= prevSum) {
            variants += calculateVariants(k + 1, preffixSum, length);
        }
    }
    
    MEMOIZED_VALUES[left][prevSum] = variants;
    IS_MEMOIZED[left][prevSum] = 1;
    return variants;
}

int main() {
    
    scanf("%s", DIGITS);
    int length;
    int testCaseNum = 1;
    
    while(DIGITS[0] != 'b') {
        
        length = strlen(DIGITS);
        
        memset(IS_MEMOIZED, 0, sizeof(IS_MEMOIZED));
        calculateCumulativeSum(length);
        printf("%d. %ld\n", testCaseNum, calculateVariants(0, 0, length));
        
        scanf("%s", DIGITS);
        testCaseNum++;
    }
}